Render emojis from Github

charlesdurham opened the pull request:
Resolves #64

charlesdurham commented:
#64

charlesdurham commented on service-api/src/test/kotlin/io/sweers/catchup/service/api/EmojiMarkdownConverterTest.kt:
```diff
@@ -0,0 +1,93 @@
+
+  // Does this scenario matter, think we'd need to move away from `Regex(":(\\w+):")` to solve.
+  @Test
+  fun FAILING_ONE_REPLACE() {
```
Note

hzsweers replied:
As in this test doesn't currently pass and we're wondering if it should?

charlesdurham replied:
From what I can tell, this test won't pass with any `Regex`. The only solution I can think of is a simple string search that looks for `:` and see's if there's an emoji alias.

Left it in since I had already written the `Replacer` before realizing it fails on this scenario and wanted to get your opinion on the cleanest solution.

charlesdurham replied:
One possible solution

```
fun replaceMarkdownEmojis(markdown: String, converter: EmojiMarkdownConverter): String {
  val sb = StringBuilder(markdown.length)
  var potentialAliasStart: Int? = null

  markdown.forEachIndexed { index, char ->
    if (char == ':') {
      potentialAliasStart = if (potentialAliasStart == null) {
        // If we have no potential start, any : is a potential start
        index
      } else {
        val potentialAlias = markdown.substring(potentialAliasStart!!, index + 1)
        val potentialEmoji = converter.convert(potentialAlias)
        // If we find an emoji append it and reset alias start, if we don't find an emoji
        // append between the potential start and this index *and* consider this index the new
        // potential start.
        if (potentialEmoji != null) {
          sb.append(potentialEmoji)
          null
        } else {
          sb.append(markdown, potentialAliasStart!!, index)
          index
        }
      }
      // While not looking for an alias end append all non possible alias chars to the string
    } else if (potentialAliasStart == null) {
      sb.append(char)
    }
  }

  // Finished iterating markdown while looking for an end, append anything remaining
  if (potentialAliasStart != null) {
    sb.append(markdown, potentialAliasStart!!, markdown.length)
  }

  return sb.toString()
}
```

charlesdurham commented on app/src/main/kotlin/io/sweers/catchup/data/gemoji/Gemoji.kt:
```diff
@@ -0,0 +1,34 @@
+data class Gemoji(
+    @PrimaryKey
+    @ColumnInfo(name = GEMOJI_ALIAS_COLUMN_NAME)
+    var alias: String,
```
Yeah, it's not necessary to do it this way. I wanted to have `alias` as the primary key. It would probably make some of the transformation code from the `GemojiJson` -> `Gemoji` simpler if I kept it as emoji to list of alias. 

Let me know if you want me to make that change.

charlesdurham commented on app/src/main/kotlin/io/sweers/catchup/data/gemoji/GemojiDatabase.kt:
```diff
@@ -0,0 +1,93 @@
+
+  return Room.databaseBuilder(context, GemojiDatabase::class.java, name)
+      .apply { builder?.invoke(this) }
+      .addCallback(object: RoomDatabase.Callback() {
```
Initially I thought the answer was no but maybe, I'll have to investigate https://issuetracker.google.com/issues/62185732

My current implementation is based off of an adaptation of tip 1 from: https://medium.com/google-developers/7-pro-tips-for-room-fbadea4bfbd1

charlesdurham commented on app/src/main/kotlin/io/sweers/catchup/data/gemoji/GemojiDatabase.kt:
```diff
@@ -0,0 +1,93 @@
+import io.sweers.catchup.R
+import okio.Okio
+
+fun GemojiDatabase(context: Context, name: String, moshi: Moshi,
```
Not sure what you mean, do you mean change it from a fake function constructor to an actual constructor?

hzsweers replied:
oh I missed that this was a `fun`, nevermind! 🙈

charlesdurham replied:
Yeah, I personally like the syntax but if you prefer something like `createGemojiDatabase(...)` let me know and I'll change it.

hzsweers replied:
yeah I'd prefer that. Upper camel makes me think it's a type

charlesdurham commented:
Changed a few things:

1. Created a module for `gemoji`
2. Fixed the failing test which required moving away from the simple regex 😞 let me know if you've got any other ideas.
3. Made `EmojiMarkdownConverter` an interface instead of function type alias.

charlesdurham commented:
I was able to pre-package the DB by slightly adapting https://issuetracker.google.com/issues/62185732#comment6. There's some migration stuff in that implementation that doesn't play nicely with Room. My "migration" is to re-copy the DB from `/assets` on a version change and tell room it was created, not updated.

hzsweers commented on gemoji/src/main/kotlin/io/sweers/catchup/gemoji/EmojiMarkdownConverter.kt:
```diff
@@ -0,0 +1,74 @@
+  val sb = StringBuilder(markdown.length)
+  var potentialAliasStart: Int? = null
+
+  markdown.forEachIndexed { index, char ->
```
TIL

hzsweers commented on gemoji/src/main/kotlin/io/sweers/catchup/gemoji/EmojiMarkdownConverter.kt:
```diff
@@ -0,0 +1,74 @@
+ */
+fun replaceMarkdownEmojis(markdown: String, converter: EmojiMarkdownConverter): String {
+  val sb = StringBuilder(markdown.length)
+  var potentialAliasStart: Int? = null
```
nit: since we're bangbanging it below already and not getting the nullability benefit, should we just make this a regular Int and use -1 as a NO_INDEX state?

hzsweers commented on scripts/gemoji/gemoji.py:
```diff
@@ -0,0 +1,43 @@
+#
```
nit: can you add the shebang here?

hzsweers replied:
also - how did we as an engineer culture arrive at that word to describe this pattern

hzsweers commented on scripts/gemoji/gemoji.py:
```diff
@@ -0,0 +1,43 @@
+# limitations under the License.
+#
+
+import sqlite3
```
neat.gif

hzsweers commented on gemoji/src/main/kotlin/io/sweers/catchup/gemoji/AssetSQLiteOpenHelper.kt:
```diff
@@ -16,92 +15,113 @@
-  init {
-    setForcedUpgrade()
+    var success = false
+    var db: SQLiteDatabase? = null
```
Any reason not to use `use` here?

charlesdurham replied:
We only close the `db` if we were unsuccessful for some reason. The other close is on the `database` field. Tried to improve the readability with https://github.com/hzsweers/CatchUp/pull/96/commits/5f39faa99b1b44cfcd8185808a589517622eb559

hzsweers commented on gemoji/src/main/kotlin/io/sweers/catchup/gemoji/SupportAssetSQLiteOpenHelper.kt:
```diff
@@ -0,0 +1,98 @@
+ * limitations under the License.
+ */
+
+// Need package to access FrameworkSQLiteDatabase
```
Could you add a longer doc here detailing this file a bit?

hzsweers commented on app/src/main/kotlin/io/sweers/catchup/data/GithubEmojiModule.kt:
```diff
@@ -0,0 +1,43 @@
+import javax.inject.Singleton
+
+@Module
+internal object GithubEmojiModule {
```
Could this be moved to the gemoji module too?

charlesdurham commented:
I've addressed most of the comments, struggling a bit with:
> Could this be moved to the gemoji module too?

Yes and it seems like it should butttttt ....

The gemoji dagger module needs a context that isn't being resolved when I move it to the gemoji gradle module because it's not qualified with `@ApplicationContext` since that's currently defined in the app gradle module. Not really sure what the best solution is, I could move the qualifier annotations to util then depend them in the gemoji gradle module?

charlesdurham commented on gemoji/src/main/kotlin/io/sweers/catchup/gemoji/GemojiDatabase.kt:
```diff
@@ -22,7 +22,7 @@ import android.arch.persistence.room.Room
 import android.content.Context
 
-fun GemojiDatabase(context: Context, name: String): GemojiDatabase {
+fun createGemojiDatabase(context: Context, name: String): GemojiDatabase {
```
I think it makes sense to remove this function and just inline it inside of `provideGemojiDatabase`, waiting until the dagger module gets moved.

hzsweers commented:
> I could move the qualifier annotations to util then depend them in the gemoji gradle module?

Yep that seems reasonable to me

hzsweers commented:
Boom! 🎉

hzsweers commented:
This has been a fantastic addition, and I learned a lot in the process. Thanks a ton for going through the iterations with me on review and trying different approaches. Really happy with the final result, and I'll get the new release out this weekend!

efemoney commented:
Loved every bit of this too. Nice work @charlesdurham @hzsweers
